lib/puppet/type/file/content_uri.rb

```diff
@@ -0,0 +1,42 @@
+require 'net/http'
+require 'uri'
+
+module Puppet
+  Puppet::Type.type(:file).newparam(:content_uri) do
+
+    desc <<-'EOT'
+      A unique identifier for file content, which must be a URI which points
+      to a file in a module. This parameter allows the content of a file resource
+      to remain consistent regardless of code updates on the master.
+
+      The normal form of a `puppet:` URI is:
+
+      `puppet:///modules/<MODULE NAME>/files/<FILE PATH>`
+
+      Note that content_uri contains "files" in its path, unlike the source parameter.
+    EOT
+
+    validate do |path|
+      begin
+        uri = URI.parse(URI.escape(path))
+      rescue
+        self.fail Puppet::Error, "Could not understand URI #{path}: #{detail}", detail
+      end
+
+      self.fail "Cannot use opaque URLs '#{source}'" unless uri.hierarchical?
+      unless uri.scheme == "puppet"
+        self.fail "Must use URLs of type puppet as content URI"
+      end
+    end
+
+    def get_from_puppet_source(source_or_content, &block)
+      request = Puppet::Indirector::Request.new(:code_content, :find, resource[:content_uri], nil, :environment => resource.catalog.environment_instance)
```

Comment:
The call below to `connection.request_get` will make the `GET` request and should raise a ruby HTTP exception. I assume we'd only make a `code_content` request if we have a `content_uri` parameter in the catalog, which we must have gotten from a new master. There is a chance you could have a cached catalog from a newer master, but point the agent to an older master. But at that point, the agent is going to need a new catalog.

:+1: to adding a spec test using `vcr`. I'd also be curious how the 404 exception propagates out, and at what point is it caught and logged. Presumably, it would cause the `resource` to fail with a meaningful error message. That said there is also a ticket to investigate network compatibility for this sort of thing, so I'd be ok deferring that work until later (and not blocking this PR).